mlflow/types/schema.py

```diff
@@ -116,84 +116,189 @@ def __repr__(self) -> str:
+                "is not allowed. Got column names {}".format([x.name for x in dataset])
             )
-        self._cols = cols
+        if not (
+            all(map(lambda x: isinstance(x, TensorSpec), dataset))
+            or all(map(lambda x: isinstance(x, ColSpec), dataset))
+        ):
+            raise MlflowException(
+                "Creating Schema with a combination of {0} and {1} is not supported. "
+                "Please choose one of {0} or {1}".format(ColSpec.__class__, TensorSpec.__class__)
+            )
+        self._data_rep = dataset
 
     @property
-    def columns(self) -> List[ColSpec]:
-        """The list of columns that defines this schema."""
-        return self._cols
+    def data_rep(self) -> List[ColSpec]:
+        """Representation of a dataset that defines this schema."""
+        return self._data_rep
+
+    def is_tensor_spec(self) -> bool:
+        """Return true iff this schema is specified using TensorSpec"""
+        return self.data_rep and isinstance(self.data_rep[0], TensorSpec)
 
     def column_names(self) -> List[Union[str, int]]:
-        """Get list of column names or range of indices if the schema has no column names."""
-        return [x.name or i for i, x in enumerate(self.columns)]
+        """Get list of data names or range of indices if the schema has no names."""
+        return [x.name or i for i, x in enumerate(self.data_rep)]
 
     def has_column_names(self) -> bool:
-        """ Return true iff this schema declares column names, false otherwise. """
-        return self.columns and self.columns[0].name is not None
+        """Return true iff this schema declares names, false otherwise. """
+        return self.data_rep and self.data_rep[0].name is not None
 
     def column_types(self) -> List[DataType]:
-        """ Get column types of the columns in the dataset."""
-        return [x.type for x in self._cols]
+        if self.is_tensor_spec():
+            raise MlflowException("TensorSpec only supports numpy types, use numpy_types() instead")
+        """ Get types of the represented dataset"""
+        return [x.type for x in self.data_rep]
 
     def numpy_types(self) -> List[np.dtype]:
-        """ Convenience shortcut to get the datatypes as numpy types."""
-        return [x.type.to_numpy() for x in self.columns]
+        """ Convenience shortcut to get the datatypes as numpy types. Unsupported by TensorSpec."""
+        if self.is_tensor_spec():
+            return [x.type for x in self.data_rep]
+        return [x.type.to_numpy() for x in self.data_rep]
 
     def pandas_types(self) -> List[np.dtype]:
-        """ Convenience shortcut to get the datatypes as pandas types."""
-        return [x.type.to_pandas() for x in self.columns]
+        """ Convenience shortcut to get the datatypes as pandas types. Unsupported by TensorSpec."""
+        if self.is_tensor_spec():
+            raise MlflowException("Datatype conversion is not supported by TensorSpec")
```

Comment:
We should unify the error message with column_types, e.g. `raise MlflowException("TensorSpec only supports numpy types, use numpy_types() instead")`